File: frontend/cockpit/assets/js/governed_operational_topology_continuity.js
(function () {
  "use strict";

  const PANEL_ID = "governed-operational-topology-continuity-panel";

  function text(value, fallback) {
    if (value === null || value === undefined || value === "") return fallback || "unknown";
    return String(value);
  }

  function computeContinuity(payload) {
    const multiWorkspace = payload && payload.governed_multi_workspace_orchestration ? payload.governed_multi_workspace_orchestration : {};
    const workspace = payload && payload.governed_runtime_workspace_continuity ? payload.governed_runtime_workspace_continuity : {};
    const orchestration = payload && payload.governed_operational_session_orchestration ? payload.governed_operational_session_orchestration : {};
    const cohesion = payload && payload.multi_panel_runtime_cohesion ? payload.multi_panel_runtime_cohesion : {};
    const loop = payload && payload.continuous_browser_runtime_loop ? payload.continuous_browser_runtime_loop : {};

    const topologySummary = multiWorkspace.summary || {};
    const workspaceSummary = workspace.summary || {};
    const sessionSummary = orchestration.summary || {};
    const cohesionSummary = cohesion.summary || {};
    const loopSummary = loop.summary || {};

    const workspaceTotal = Number(topologySummary.workspace_total || 0);
    const missingTotal = Number(topologySummary.missing_panel_total || cohesionSummary.missing_total || 0);
    const driftTotal = Number(topologySummary.authority_drift_total || cohesionSummary.drift_total || 0);

    const topologyState = multiWorkspace.topology_state || "unknown";
    const workspaceState = workspace.workspace_state || "unknown";
    const sessionState = orchestration.session_state || "unknown";
    const loopState = loop.loop_state || "unknown";

    let continuityState = "continuous";
    if (driftTotal || topologyState === "blocked") continuityState = "blocked";
    else if (missingTotal || topologyState === "partial") continuityState = "partial";
    else if (topologyState === "degraded" || topologyState === "recovering") continuityState = topologyState;
    else if (topologyState === "review_required" || workspaceState === "review_required" || sessionState === "review_required") continuityState = "review_required";
    else if (topologyState === "unknown" && workspaceState === "unknown") continuityState = "initializing";

    return {
      version: "v19.89.8-S31",
      status: "active",
      continuity_state: continuityState,
      authority: {
        backend_owns_truth: true,
        cockpit_presentation_only: true,
        runtime_authority: "blocked",
        browser_execution_authority: "blocked",
        operator_mutation_enabled: false,
        runtime_mutation_enabled: false,
        workspace_mutation_enabled: false,
        topology_mutation_enabled: false,
        fail_closed_governance: true,
        autonomous_execution_expansion: false
      },
      summary: {
        workspace_total: workspaceTotal,
        selected_route: topologySummary.selected_route || workspaceSummary.selected_route || "unknown",
        topology_state: topologyState,
        workspace_state: workspaceState,
        session_state: sessionState,
        loop_state: loopState,
        missing_panel_total: missingTotal,
        authority_drift_total: driftTotal,
        payload_propagation: topologySummary.payload_propagation || cohesionSummary.payload_propagation || "unknown",
        manual_review_required: Boolean(topologySummary.manual_review_required || workspaceSummary.manual_review_required)
      },
      continuity_chain: [
        { layer: "multi_workspace_topology", state: topologyState, detail: workspaceTotal + " workspaces" },
        { layer: "workspace_continuity", state: workspaceState, detail: workspaceSummary.selected_route || "unknown" },
        { layer: "session_orchestration", state: sessionState, detail: sessionSummary.selected_route || "unknown" },
        { layer: "panel_cohesion", state: cohesion.cohesion_state || "unknown", detail: cohesionSummary.payload_propagation || "unknown" },
        { layer: "browser_loop", state: loopState, detail: loopSummary.payload_freshness || "unknown" }
      ]
    };
  }

  function stateClass(state) {
    if (state === "blocked") return "topology-continuity-blocked";
    if (state === "partial") return "topology-continuity-partial";
    if (state === "review_required") return "topology-continuity-review";
    if (state === "degraded") return "topology-continuity-degraded";
    if (state === "recovering") return "topology-continuity-recovering";
    if (state === "initializing") return "topology-continuity-initializing";
    return "topology-continuity-continuous";
  }

  function ensurePanel() {
    let panel = document.getElementById(PANEL_ID);
    if (panel) return panel;

    panel = document.createElement("section");
    panel.id = PANEL_ID;
    panel.className = "claire-card governed-operational-topology-continuity-card";
    panel.innerHTML = `
      <div class="topology-continuity-header">
        <div>
          <div class="topology-continuity-kicker">Operational Topology Continuity</div>
          <h2>Topology Continuity</h2>
        </div>
        <div class="topology-continuity-badge">presentation only - topology mutation blocked</div>
      </div>
      <div id="topology-continuity-summary" class="topology-continuity-summary">Waiting for canonical payload.</div>
      <div id="topology-continuity-chain" class="topology-continuity-chain"></div>
    `;

    const targets = ["#runtime-surface", "#operator-surface", "#main-content", "main", "#app", "body"];
    for (const selector of targets) {
      const target = document.querySelector(selector);
      if (target) {
        target.appendChild(panel);
        return panel;
      }
    }
    document.body.appendChild(panel);
    return panel;
  }

  function render(payload) {
    const panel = ensurePanel();
    const continuity = payload && payload.governed_operational_topology_continuity
      ? payload.governed_operational_topology_continuity
      : computeContinuity(payload || {});
    const summary = continuity.summary || {};
    const chain = Array.isArray(continuity.continuity_chain) ? continuity.continuity_chain : [];
    const summaryEl = panel.querySelector("#topology-continuity-summary");
    const chainEl = panel.querySelector("#topology-continuity-chain");

    panel.classList.remove("topology-continuity-blocked", "topology-continuity-partial", "topology-continuity-review", "topology-continuity-degraded", "topology-continuity-recovering", "topology-continuity-initializing", "topology-continuity-continuous");
    panel.classList.add(stateClass(continuity.continuity_state));

    summaryEl.innerHTML = `
      <span><strong>Continuity:</strong> ${text(continuity.continuity_state, "unknown")}</span>
      <span><strong>Workspaces:</strong> ${text(summary.workspace_total, "0")}</span>
      <span><strong>Route:</strong> ${text(summary.selected_route, "unknown")}</span>
      <span><strong>Topology:</strong> ${text(summary.topology_state, "unknown")}</span>
      <span><strong>Session:</strong> ${text(summary.session_state, "unknown")}</span>
      <span><strong>Mutation:</strong> blocked</span>
    `;

    chainEl.innerHTML = chain.map((item) => `
      <div class="topology-continuity-node">
        <div class="topology-continuity-node-layer">${text(item.layer, "layer")}</div>
        <div class="topology-continuity-node-state">${text(item.state, "unknown")}</div>
        <div class="topology-continuity-node-detail">${text(item.detail, "unknown")}</div>
      </div>
    `).join("");
  }

  function poll() {
    fetch("/dashboard/payload", { cache: "no-store" })
      .then((response) => response.ok ? response.json() : null)
      .then((payload) => { if (payload) render(payload); })
      .catch(() => ensurePanel());
  }

  document.addEventListener("DOMContentLoaded", function () {
    ensurePanel();
    window.addEventListener("claire:canonical-payload", function (event) { render(event.detail || {}); });
    window.addEventListener("claire:payload", function (event) { render(event.detail || {}); });
    poll();
    setInterval(poll, 10000);
  });

  window.ClaireGovernedOperationalTopologyContinuity = {
    version: "v19.89.8-S31",
    render,
    authority: "presentation_only_runtime_authority_blocked"
  };
})();

File: claire/api/governed_operational_topology_continuity.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List


OPERATIONAL_TOPOLOGY_CONTINUITY_VERSION = "v19.89.8-S31"


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def build_operational_topology_continuity(payload: Dict[str, Any] | None = None) -> Dict[str, Any]:
    payload = payload if isinstance(payload, dict) else {}

    multi_workspace = payload.get("governed_multi_workspace_orchestration") if isinstance(payload.get("governed_multi_workspace_orchestration"), dict) else {}
    workspace = payload.get("governed_runtime_workspace_continuity") if isinstance(payload.get("governed_runtime_workspace_continuity"), dict) else {}
    orchestration = payload.get("governed_operational_session_orchestration") if isinstance(payload.get("governed_operational_session_orchestration"), dict) else {}
    cohesion = payload.get("multi_panel_runtime_cohesion") if isinstance(payload.get("multi_panel_runtime_cohesion"), dict) else {}
    loop = payload.get("continuous_browser_runtime_loop") if isinstance(payload.get("continuous_browser_runtime_loop"), dict) else {}

    topology_summary = multi_workspace.get("summary") if isinstance(multi_workspace.get("summary"), dict) else {}
    workspace_summary = workspace.get("summary") if isinstance(workspace.get("summary"), dict) else {}
    session_summary = orchestration.get("summary") if isinstance(orchestration.get("summary"), dict) else {}
    cohesion_summary = cohesion.get("summary") if isinstance(cohesion.get("summary"), dict) else {}
    loop_summary = loop.get("summary") if isinstance(loop.get("summary"), dict) else {}

    workspace_total = int(topology_summary.get("workspace_total", 0) or 0)
    missing_total = int(topology_summary.get("missing_panel_total", cohesion_summary.get("missing_total", 0)) or 0)
    drift_total = int(topology_summary.get("authority_drift_total", cohesion_summary.get("drift_total", 0)) or 0)

    topology_state = multi_workspace.get("topology_state", "unknown")
    workspace_state = workspace.get("workspace_state", "unknown")
    session_state = orchestration.get("session_state", "unknown")
    loop_state = loop.get("loop_state", "unknown")

    continuity_state = "continuous"
    if drift_total or topology_state == "blocked":
        continuity_state = "blocked"
    elif missing_total or topology_state == "partial":
        continuity_state = "partial"
    elif topology_state in {"degraded", "recovering"}:
        continuity_state = topology_state
    elif topology_state == "review_required" or workspace_state == "review_required" or session_state == "review_required":
        continuity_state = "review_required"
    elif topology_state == "unknown" and workspace_state == "unknown":
        continuity_state = "initializing"

    continuity_chain: List[Dict[str, Any]] = [
        {
            "layer": "multi_workspace_topology",
            "state": topology_state,
            "detail": f"{workspace_total} workspaces",
        },
        {
            "layer": "workspace_continuity",
            "state": workspace_state,
            "detail": workspace_summary.get("selected_route", "unknown"),
        },
        {
            "layer": "session_orchestration",
            "state": session_state,
            "detail": session_summary.get("selected_route", "unknown"),
        },
        {
            "layer": "panel_cohesion",
            "state": cohesion.get("cohesion_state", "unknown"),
            "detail": cohesion_summary.get("payload_propagation", "unknown"),
        },
        {
            "layer": "browser_loop",
            "state": loop_state,
            "detail": loop_summary.get("payload_freshness", "unknown"),
        },
    ]

    return {
        "version": OPERATIONAL_TOPOLOGY_CONTINUITY_VERSION,
        "status": "active",
        "continuity_state": continuity_state,
        "observed_at_utc": _utc_now(),
        "authority": {
            "backend_owns_truth": True,
            "cockpit_presentation_only": True,
            "runtime_authority": "blocked",
            "browser_execution_authority": "blocked",
            "operator_mutation_enabled": False,
            "runtime_mutation_enabled": False,
            "workspace_mutation_enabled": False,
            "topology_mutation_enabled": False,
            "fail_closed_governance": True,
            "autonomous_execution_expansion": False,
        },
        "summary": {
            "workspace_total": workspace_total,
            "selected_route": topology_summary.get("selected_route", workspace_summary.get("selected_route", "unknown")),
            "topology_state": topology_state,
            "workspace_state": workspace_state,
            "session_state": session_state,
            "loop_state": loop_state,
            "missing_panel_total": missing_total,
            "authority_drift_total": drift_total,
            "payload_propagation": topology_summary.get("payload_propagation", cohesion_summary.get("payload_propagation", "unknown")),
            "manual_review_required": bool(topology_summary.get("manual_review_required", False) or workspace_summary.get("manual_review_required", False)),
        },
        "continuity_chain": continuity_chain,
    }


def attach_operational_topology_continuity(payload: Dict[str, Any]) -> Dict[str, Any]:
    payload = dict(payload) if isinstance(payload, dict) else {}
    payload["governed_operational_topology_continuity"] = build_operational_topology_continuity(payload)
    return payload
